From c19d5e1338501335bb2fd0b08f6685c1cce0009e Mon Sep 17 00:00:00 2001
From: George Burgess IV <gbiv@google.com>
Date: Mon, 10 Aug 2026 10:47:47 -0600
Subject: [PATCH] Add temporary triage handler for libcxx hardening

This seeks to point people very directly to b/543356317 if
they see a libcxx hardening related crash. This is guarded
under a -D flag so as to minimize the chance that e.g., an
NDK user bumps into this.

I was hoping to use `__builtin_verbose_trap` support,
but that embeds messages entirely in debuginfo, so
tombstones don't directly surface the 'helpful' message;
seems we need an actual frame for that to work, so
a `noinline` function is required.

The intent is to drop this relatively quickly (O(weeks),
rather than carrying it for months).

Bug: 543356317
Test: mma
Change-Id: I2e6356c62416f611f11391bb92336ea04f19d026
---
 libcxx/include/__verbose_abort                    | 15 +++++++++++++++
 .../vendor/android/android_assertion_handler.in   |  5 ++++-
 libcxx/vendor/llvm/default_assertion_handler.in   |  3 +++
 3 files changed, 22 insertions(+), 1 deletion(-)

diff --git a/libcxx/include/__verbose_abort b/libcxx/include/__verbose_abort
index f8b6967..18c5e7c 100644
--- a/libcxx/include/__verbose_abort
+++ b/libcxx/include/__verbose_abort
@@ -23,6 +23,21 @@ _LIBCPP_BEGIN_NAMESPACE_STD
 [[__noreturn__]] _LIBCPP_AVAILABILITY_VERBOSE_ABORT _LIBCPP_OVERRIDABLE_FUNC_VIS _LIBCPP_ATTRIBUTE_FORMAT(
     __printf__, 1, 2) void __libcpp_verbose_abort(const char* __format, ...) _NOEXCEPT;
 
+// NOTE(b/543356317): This is only intended to be used as we 'shake out'
+// initial libcxx hardening crashes from Android (a few weeks). It's not called
+// by a stable hardening mode, and not planned to be upstreamed. The intent is
+// _entirely_ to point people to the libcxx hardening enablement metabug in
+// backtraces.
+[[__noreturn__]] [[gnu::noinline]] [[clang::nomerge]] inline void __libcpp_hardening_failure_PLEASE_SEE_BUG_543356317(const char *) {
+  // Store to a volatile to defeat ICF.
+  [[maybe_unused]] volatile unsigned i = 0x1234567;
+  i = 0x89ABC;
+  __builtin_trap();
+}
+
+#define _ANDROID_LIBCPP_TRIAGE_VERBOSE_ABORT(message) \
+  ::std::__libcpp_hardening_failure_PLEASE_SEE_BUG_543356317(message)
+
 // _LIBCPP_VERBOSE_ABORT(format, args...)
 //
 // This macro is used to abort the program abnormally while providing additional diagnostic information.
diff --git a/libcxx/vendor/android/android_assertion_handler.in b/libcxx/vendor/android/android_assertion_handler.in
index d48036b..c2230a1 100644
--- a/libcxx/vendor/android/android_assertion_handler.in
+++ b/libcxx/vendor/android/android_assertion_handler.in
@@ -19,7 +19,10 @@
 #  pragma GCC system_header
 #endif
 
-#if _LIBCPP_ASSERTION_SEMANTIC == _LIBCPP_ASSERTION_SEMANTIC_IGNORE
+#if defined(_ANDROID_LIBCPP_TRIAGE_HANDLER)
+#  define _LIBCPP_ASSERTION_HANDLER(message) _ANDROID_LIBCPP_TRIAGE_VERBOSE_ABORT(message)
+
+#elif _LIBCPP_ASSERTION_SEMANTIC == _LIBCPP_ASSERTION_SEMANTIC_IGNORE
 #  define _LIBCPP_ASSERTION_HANDLER(message) ((void)0)
 
 #elif _LIBCPP_ASSERTION_SEMANTIC == _LIBCPP_ASSERTION_SEMANTIC_OBSERVE
diff --git a/libcxx/vendor/llvm/default_assertion_handler.in b/libcxx/vendor/llvm/default_assertion_handler.in
index d352405..f5ce288 100644
--- a/libcxx/vendor/llvm/default_assertion_handler.in
+++ b/libcxx/vendor/llvm/default_assertion_handler.in
@@ -35,6 +35,9 @@
 #    define _LIBCPP_ASSERTION_HANDLER(message) _LIBCPP_VERBOSE_TRAP(message)
 #  endif // _LIBCPP_HARDENING_MODE == _LIBCPP_HARDENING_MODE_DEBUG
 
+#elif defined(_ANDROID_LIBCPP_TRIAGE_HANDLER)
+#    define _LIBCPP_ASSERTION_HANDLER(message) _ANDROID_LIBCPP_TRIAGE_VERBOSE_ABORT(message)
+
 #else
 
 #  if _LIBCPP_ASSERTION_SEMANTIC == _LIBCPP_ASSERTION_SEMANTIC_IGNORE
-- 
2.55.0.1003.g10538fe699-goog

